storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +63,403 @@
+    private final long pollTimeoutMs;
     private final Time time;
 
-    // It indicates whether the closing process has been started or not. If it is set as true,
-    // consumer will stop consuming messages, and it will not allow partition assignments to be updated.
-    private volatile boolean closing = false;
-
-    // It indicates whether the consumer needs to assign the partitions or not. This is set when it is
-    // determined that the consumer needs to be assigned with the updated partitions.
-    private volatile boolean assignPartitions = false;
+    // It indicates whether the ConsumerTask is closed or not.
+    private volatile boolean isClosed = false;
+    // It indicates whether the user topic partition assignment to the consumer has changed or not. If the assignment
+    // has changed, the consumer will eventually start tracking the newly assigned partitions and stop tracking the
+    // ones it is no longer assigned to.
+    // The initial value is set to true to wait for partition assignment on the first execution; otherwise thread will
+    // be busy without actually doing anything
+    private volatile boolean hasAssignmentChanged = true;
 
     // It represents a lock for any operations related to the assignedTopicPartitions.
     private final Object assignPartitionsLock = new Object();
 
     // Remote log metadata topic partitions that consumer is assigned to.
-    private volatile Set<Integer> assignedMetaPartitions = Collections.emptySet();
+    private volatile Set<Integer> assignedMetadataPartitions = Collections.emptySet();
 
     // User topic partitions that this broker is a leader/follower for.
-    private Set<TopicIdPartition> assignedTopicPartitions = Collections.emptySet();
+    private volatile Map<TopicIdPartition, UserTopicIdPartition> assignedUserTopicIdPartitions = Collections.emptyMap();
+    private volatile Set<TopicIdPartition> processedAssignmentOfUserTopicIdPartitions = Collections.emptySet();
 
-    // Map of remote log metadata topic partition to consumed offsets. Received consumer records
-    // may or may not have been processed based on the assigned topic partitions.
-    private final Map<Integer, Long> partitionToConsumedOffsets = new ConcurrentHashMap<>();
+    private long uninitializedAt;
+    private boolean isAllUserTopicPartitionsInitialized;
 
-    // Map of remote log metadata topic partition to processed offsets that were synced in committedOffsetsFile.
-    private Map<Integer, Long> lastSyncedPartitionToConsumedOffsets = Collections.emptyMap();
+    // Map of remote log metadata topic partition to consumed offsets.
+    private final Map<Integer, Long> readOffsetsByMetadataPartition = new ConcurrentHashMap<>();
+    private final Map<TopicIdPartition, Long> readOffsetsByUserTopicPartition = new HashMap<>();
 
-    private final long committedOffsetSyncIntervalMs;
-    private CommittedOffsetsFile committedOffsetsFile;
```

Comment:
@abhijeetk88 @satishd : Sorry to chime in late on this. It seems that we removed the usage of `CommittedOffsetsFile`.
1. Does the consumer always start from the beginning offset after restart?
2. Should we remove the CommittedOffsetsFile class?
3. This actually changes the on-disk layout. Was this discussed in a KIP?

Reply:
@junrao , answering your question:

    1. Does the consumer always start from the beginning offset after restart?
    Yes, we always read from beginning each time we restart.
    2. Should we remove the CommittedOffsetsFile class?
    Good point. Will remove it in another PR.
    3. This actually changes the on-disk layout. Was this discussed in a KIP?
    I had a look, and yes, the committed offset file is mentioned in the [KIP-405](https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage#KIP405:KafkaTieredStorage-Committedoffsetsfileformat). @kamalcph @satishd  , do we remove the offset committed file and read from beginning each time on purpose? To fix the original bug, I don't think we need to remove this offset committed file mechanism. WDYT?

Reply:
Sorry for the late reply. Now, we are always reading from the beginning of the topic to restore the state. And, the committed offsets design is unused and kept as-is for future improvement. Storing the snapshot of the remote log metadata under each partition requires some discussion.

How often to trigger the snapshot?

1. If we configure to run the snapshots in a scheduled interval, then it can lead to flushing of the log segments which is async now.
2. On broker shutdown: This can lead to two cases: clean and unclean shutdown. During unclean shutdown, we cannot depend on the snapshot stored in the local disk as it can be corrupted. Reading from the source-of-truth (__remote_log_metadata topic) is reliable (v0).
3. On Partition reassignments/New-Topic: When a user-topic-partition (utp) is reassigned to another broker, it has to reload the context from the `__remote_log_metadata` topic.
     
For V0, we kept the design to read from the beginning of the respective metadata topic-partition.

Reply:
Sorry @junrao for missing this comment in my list of notifications. 

Adding to what @showuon and @kamalcph mentioned above.

>Should we remove the CommittedOffsetsFile class?
We may want to keep it in the repo and use it in the future or we can remove it for now and add it back when we work on this in future. I am fine with either way.

Reply:
@kamalcph @satishd @showuon : Thanks for the reply. Do we plan to use `CommittedOffsetsFile` anytime soon? If not, it's probably better to remove it for now and add it back when it's being used. Also, have we decided if `CommittedOffsetsFile` will be used in the same way as described in KIP-405? If not, it would be useful to update the KIP with the new design and send the update to the vote thread.

Reply:
@junrao We will remove the CommittedOffsetsFile for now as you suggested. The current plan is to use it as mentioned in KIP-405. We will follow KIP update process and notify in the vote thread if it is different from the existing proposal.